### PR TITLE
Support high water mark temporal flame graphs

### PR DESCRIPTION
**Pull request by godlygeek**
Building on top of the work to support our temporal flame graphs for analyzing memory leaked within a user-selected range, allow creating temporal flame graphs for analyzing the high water mark within a user-selected range.

### CODE REVIEW

**Comment by godlygeek** [comment]
@pablogsal I'd appreciate it if you could skim through this sometime soon and share your initial impressions. I'm a bit worried that you'll be unhappy with how complex `TemporalHighWaterMarkAggregator` needs to be, so I'd welcome any suggestions for how to simplify it.

But hey, at least this one's only about 1000 lines!

**Comment by pablogsal** [comment]
> But hey, at least this one's only about 1000 lines!

Screenshot 2023-03-09 at 14 36 23

🤔

**Comment by pablogsal** [comment]
image

**Comment by godlygeek** [comment]
More than 400 LoC out of that are new tests ☹️

Let's make some time to chat about what to do with this. I'd rather not release temporal leaks without also releasing temporal high water mark reporting... And given #382 we probably should cut a release soon. I'm tempted to back out the temporal leaks stuff (at least the CLI entry point for it and any docs) until we're comfortable landing this.

I'm gonna rebase this, and also take another pass and see if I can find any ways to simplify it.

**Comment by pablogsal** [comment]
@godlygeek You can rebase this now :)

**Comment by godlygeek** [comment]
Done. I'm still tweaking this a bit, so I'm going to leave it as draft, but I'd love it if you could take another look at the first commit, in particular, and see if there's anything that stands out to you as wrong or bad or difficult to understand.

I think I need to add more comments, and maybe some diagrams, but the implementation should remain basically unchanged

File: src/memray/_memray/snapshot.cpp [hunk 1]
```diff
@@ -308,9 +320,61 @@ UsageHistory::UsageHistoryImpl::rebase(size_t new_peak)
...
+        std::vector<HistoricalContribution>& heap_contribution_by_snapshot) const
+{
+    size_t current_snapshot = highest_peak_by_snapshot.size();
+    auto history = d_history;
+    while (history.last_known_snapshot != current_snapshot) {
+        assert(history.last_known_snapshot < current_snapshot);
+        size_t last_snapshot_peak = highest_peak_by_snapshot.at(history.last_known_snapshot);
+        history.rebase(last_snapshot_peak);
+
+        HistoricalContribution hc;
```
Comment (pablogsal) [inline src/memray/_memray/snapshot.cpp:335]: Why not using the brace-constructor here?

File: src/memray/_memray/snapshot.cpp [hunk 2]
```diff
@@ -341,6 +422,51 @@ UsageHistory::leaksContribution(size_t current_peak) const
...
+        hc.contrib = hwm;
+        ret.push_back(hc);
+    }
+
+    if (ret.empty() || ret.back().contrib != leaks) {
+        HistoricalContribution hc;
+        hc.as_of_snapshot = final.last_known_snapshot + 1;
+        hc.peak_index = -1;
+        hc.contrib = leaks;
+        ret.push_back(hc);
```
Comment (pablogsal) [inline src/memray/_memray/snapshot.cpp:464]: Same as before: We can create the element in-pace using `emplace_back`

**Review by pablogsal — APPROVED** [review]
LGTM

**Pull request merged**